src/3d/qgspointcloudlayerchunkloader_p.cpp

```diff
@@ -255,4 +256,107 @@ QgsPointCloudLayerChunkedEntity::~QgsPointCloudLayerChunkedEntity()
+  QgsVector3D directionMapCoords = pointMapCoords - originMapCoords;
+  directionMapCoords.normalize();
+
+  const QVector3D rayOriginMapCoords( originMapCoords.x(), originMapCoords.y(), originMapCoords.z() );
+  const QVector3D rayDirectionMapCoords( directionMapCoords.x(), directionMapCoords.y(), directionMapCoords.z() );
+
+  const int screenSizePx = std::max( context.screenSize.height(), context.screenSize.width() );
+
+  const QgsPointCloud3DSymbol *symbol = factory->mSymbol.get();
+  // Symbol can be null in case of no rendering enabled
+  if ( !symbol )
+    return result;
+  const double pointSize = symbol->pointSize();
+
+  const double limitAngle = 2 * pointSize / screenSizePx * factory->mMap.fieldOfView();
+
+  // adjust ray to elevation properties
+  const QVector3D adjutedRayOrigin = QVector3D( rayOriginMapCoords.x(), rayOriginMapCoords.y(), ( rayOriginMapCoords.z() -  factory->mZValueOffset ) / factory->mZValueScale );
+  QVector3D adjutedRayDirection = QVector3D( rayDirectionMapCoords.x(), rayDirectionMapCoords.y(), rayDirectionMapCoords.z() / factory->mZValueScale );
+  adjutedRayDirection.normalize();
+
+  const QgsRay3D layerRay( adjutedRayOrigin, adjutedRayDirection );
+
+  QgsPointCloudIndex *index = factory->mPointCloudIndex;
+
+  const QgsPointCloudAttributeCollection attributeCollection = index->attributes();
+  QgsPointCloudRequest request;
+  request.setAttributes( attributeCollection );
+
+  float minDist = -1;
+  const QList<QgsChunkNode *> activeNodes = this->activeNodes();
+  for ( QgsChunkNode *node : activeNodes )
+  {
+    const QgsChunkNodeId id = node->tileId();
+    const IndexedPointCloudNode n( id.d, id.x, id.y, id.z );
+
+    if ( !index->hasNode( n ) )
+      continue;
+
+    if ( !ray.intersects( Qgs3DUtils::aabbToBox( node->bbox() ) ) )
+      continue;
+
+    std::unique_ptr<QgsPointCloudBlock> block( index->nodeData( n, request ) );
+    if ( !block )
+      continue;
+
+    const QgsVector3D blockScale = block->scale();
+    const QgsVector3D blockOffset = block->offset();
+
+    const char *ptr = block->data();
+    const QgsPointCloudAttributeCollection blockAttributes = block->attributes();
+    const std::size_t recordSize = blockAttributes.pointRecordSize();
+    int xOffset = 0, yOffset = 0, zOffset = 0;
+    const QgsPointCloudAttribute::DataType xType = blockAttributes.find( QStringLiteral( "X" ), xOffset )->type();
+    const QgsPointCloudAttribute::DataType yType = blockAttributes.find( QStringLiteral( "Y" ), yOffset )->type();
+    const QgsPointCloudAttribute::DataType zType = blockAttributes.find( QStringLiteral( "Z" ), zOffset )->type();
+    for ( int i = 0; i < block->pointCount(); ++i )
+    {
+      double x, y, z;
+      QgsPointCloudAttribute::getPointXYZ( ptr, i, recordSize, xOffset, xType, yOffset, yType, zOffset, zType, blockScale, blockOffset, x, y, z );
+      const QVector3D point( x, y, z );
```

Comment:
I think this is going to cause trouble due to double->float conversion. I think we should transform the points to world coordinates first, and only then convert the "small" numbers to floats - and ray should also stay in world coordinates for numerical stability.